src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs

```diff
@@ -608,103 +615,124 @@ private async Task ProcessRequests<TContext>(IHttpApplication<TContext> applicat
 
                 InitializeBodyControl(messageBody);
 
-                var context = application.CreateContext(this);
+                // Set the TContext for the request callback
+                requestData.Context = application.CreateContext(this);
 
-                try
-                {
-                    KestrelEventSource.Log.RequestStart(this);
+                // Start each request on a fresh ExecutionContext
+                ExecutionContext.Run(cleanExecutionContext, contextCallback, requestData);
+                // Await the request Task set in the callback; then clear it
+                await requestData.CurrentRequest;
+                requestData.CurrentRequest = default;
 
-                    // Run the application code for this request
-                    await application.ProcessRequestAsync(context);
+                // Clear the TContext for the request
+                application.DisposeContext(requestData.Context, _applicationException);
```

Comment:
Given that it's pluggable, should CreateContext and DisposeContext be called with the `cleanExecutionContext`?

Reply:
I suppose

Reply:
Done